zulu/datetime.py

```diff
@@ -379,6 +380,154 @@ def replace(self,
+
+    def start_of_second(self):
+        """This method helps in shifting the values to start of the second."""
+        return self.shift(microseconds=-self.microsecond)
+
+    def end_of_year(self, count):
+        """This method helps in shifting the values to end of the year."""
+        return self.shift(years=count * 1,
+                          months=-self.month + 1,
+                          days=-self.day + 1,
+                          hours=-self.hour,
+                          minutes=-self.minute,
+                          seconds=-self.second,
+                          microseconds=-1)
+
+    def end_of_month(self, count):
+        """This method helps in shifting the values to end of the month."""
+        return self.shift(months=count * 1,
+                          days=-self.day + 1,
+                          hours=-self.hour,
+                          minutes=-self.minute,
+                          seconds=-self.second,
+                          microseconds=-1
+                          )
+
+    def end_of_day(self, count):
+        """This method helps in shifting the values to end of the day."""
+        return self.shift(days=count * 1,
+                          hours=-self.hour,
+                          minutes=-self.minute,
+                          seconds=-self.second,
+                          microseconds=-1
+                          )
+
+    def end_of_hour(self, count):
+        """This method helps in shifting the values to end of the hour."""
+        return self.shift(hours=count * 1,
+                          minutes=-self.minute,
+                          seconds=-self.second,
+                          microseconds=-1)
+
+    def end_of_minute(self, count):
+        """This method helps in shifting the values to end of the minute."""
+        return self.shift(minutes=count * 1,
+                          seconds=-self.second,
+                          microseconds=-1)
+
+    def end_of_second(self, count):
+        """This method helps in shifting the values to end of the second."""
+        return self.shift(seconds=count * 1,
+                          microseconds=-1)
+
+    def validate_frame(self, frame):
+        """Validates the given time frame."""
+        attrs = ['year',
+                 'month',
+                 'day',
+                 'hour',
+                 'minute',
+                 'second']
+
+        if frame not in attrs:
+            raise AttributeError('The given time frame {0} is invalid'
+                                 .format(frame))
+
+    def start_of(self, frame):
+        """This method returns a floor value of datetime for the given time
+        frame.
+
+        Args:
+            frame (str): A time frame. Ex: year, month, day, minute, etc.
+
+        Returns:
+            zulu.Datetime.datetime
+        """
+        self.validate_frame(frame)
+
+        method = getattr(self, 'start_of_{0}'.format(frame), None)
+        if callable(method):
```

Comment:
Doesn't really make sense to check for callable since you're validating in `validate_frame`. I.e. if you get here, then the frame is valid and there exists a method for. If the method didn't actually exists, then the user has done something to the class and we should blow up instead of silently returning `None`.